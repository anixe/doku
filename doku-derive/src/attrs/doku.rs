mod container;
mod field;
mod variant;

pub use self::{container::*, field::*, variant::*};
