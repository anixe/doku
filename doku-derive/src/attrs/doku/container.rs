use crate::prelude::*;
use darling::FromMeta;

/// Models the `#[doku]` attribute for containers:
///
/// ```rust,ignore
/// #[doku(wrap = "container")]
/// struct Foo {
///     field: Bar,
/// }
/// ````
#[derive(Clone, Debug, Default, FromMeta)]
pub struct DokuContainer {
    #[darling(default)]
    pub content: Option<syn::LitStr>,

    #[darling(default)]
    pub tag: Option<syn::LitStr>,

    #[darling(default)]
    pub transparent: Option<bool>,

    #[darling(default)]
    pub untagged: Option<bool>,

    #[darling(default)]
    pub wrap: Option<syn::LitStr>,
}

impl DokuContainer {
    pub fn from_ast(attrs: &[syn::Attribute]) -> Result<Self> {
        attrs::from_ast(attrs, "doku").map(|attrs| attrs.fold(Self::default(), Self::merge))
    }

    fn merge(self, other: Self) -> Self {
        Self {
            content:     other.content.or(self.content),
            tag:         other.tag.or(self.tag),
            transparent: other.transparent.or(self.transparent),
            untagged:    other.untagged.or(self.untagged),
            wrap:        other.wrap.or(self.wrap),
        }
    }
}
