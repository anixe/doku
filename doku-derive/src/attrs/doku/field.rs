use crate::prelude::*;
use darling::FromMeta;

/// Models the `#[doku]` attribute for fields:
///
/// ```ignore
/// struct Foo {
///     #[doku(as = "Zar")]
///     field: Bar,
/// }
/// ````
#[derive(Clone, Debug, Default, FromMeta)]
pub struct DokuField {
    #[darling(default, rename = "as")]
    pub as_: Option<syn::LitStr>,

    #[darling(default, rename = "example", multiple)]
    pub examples: Vec<syn::LitStr>,

    #[darling(default)]
    pub flatten: Option<bool>,

    #[darling(default)]
    pub rename: Option<syn::LitStr>,

    #[darling(default)]
    pub skip: Option<bool>,

    #[darling(default)]
    pub tag: Option<syn::LitStr>,
}

impl DokuField {
    pub fn from_ast(attrs: &[syn::Attribute]) -> Result<Self> {
        attrs::from_ast(attrs, "doku")
            .map(|attrs| attrs.fold(Self::default(), Self::merge))
    }

    fn merge(self, other: Self) -> Self {
        let examples =
            self.examples.into_iter().chain(other.examples).collect();

        Self {
            as_: other.as_.or(self.as_),
            examples,
            flatten: other.flatten.or(self.flatten),
            rename: other.rename.or(self.rename),
            skip: other.skip.or(self.skip),
            tag: other.tag.or(self.tag),
        }
    }
}
