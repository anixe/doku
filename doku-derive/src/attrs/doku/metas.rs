use darling::{Error, FromMeta, Result};
use std::{collections::BTreeMap, iter::FromIterator};

const ERROR_EXPECTED_KV_LITERAL: &str =
    "Expected a key-value string literal such as: `#[doku(meta(\"key = value\"))]`";

#[derive(Clone, Debug, Default)]
pub struct DokuMetas {
    pub metas: BTreeMap<String, String>,
}

impl FromMeta for DokuMetas {
    fn from_list(items: &[syn::NestedMeta]) -> Result<Self> {
        let metas = items
            .iter()
            .map(|item| match item {
                syn::NestedMeta::Lit(syn::Lit::Str(lit)) => {
                    Ok((lit, lit.value()))
                }

                _ => {
                    Err(Error::custom(ERROR_EXPECTED_KV_LITERAL)
                        .with_span(item))
                }
            })
            .map(|item| {
                item.and_then(|(lit, lit_value)| {
                    let mut kv = lit_value.splitn(2, '=');

                    let key = kv.next().ok_or_else(|| {
                        Error::custom(ERROR_EXPECTED_KV_LITERAL).with_span(lit)
                    })?;

                    let value = kv.next().ok_or_else(|| {
                        Error::custom(ERROR_EXPECTED_KV_LITERAL).with_span(lit)
                    })?;

                    Ok((key.trim().to_owned(), value.trim().to_owned()))
                })
            })
            .collect::<Result<_>>()?;

        Ok(Self { metas })
    }
}

impl FromIterator<DokuMetas> for DokuMetas {
    fn from_iter<T>(iter: T) -> Self
    where
        T: IntoIterator<Item = DokuMetas>,
    {
        Self {
            metas: iter.into_iter().flat_map(|metas| metas.metas).collect(),
        }
    }
}
