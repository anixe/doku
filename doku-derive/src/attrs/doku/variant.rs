use crate::prelude::*;
use darling::FromMeta;

/// Models the `#[doku]` attribute for variants:
///
/// ```rust,ignore
/// enum Foo {
///     #[doku(rename = "BAR")]
///     Bar,
/// }
/// ````
#[derive(Clone, Debug, Default, FromMeta)]
pub struct DokuVariant {
    #[darling(default)]
    pub rename: Option<syn::LitStr>,

    #[darling(default)]
    pub skip: Option<bool>,
}

impl DokuVariant {
    pub fn from_ast(attrs: &[syn::Attribute]) -> Result<Self> {
        attrs::from_ast(attrs, "doku").map(|attrs| attrs.fold(Self::default(), Self::merge))
    }

    fn merge(self, other: Self) -> Self {
        Self {
            rename: other.rename.or(self.rename),
            skip:   other.skip.or(self.skip),
        }
    }
}
