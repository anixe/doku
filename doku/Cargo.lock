# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
[[package]]
name = "ansi_term"
version = "0.11.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ee49baf6cb617b853aa8d93bf420db2383fab46d314482ca2803b40d5fde979b"
dependencies = [
 "winapi",
]

[[package]]
name = "autocfg"
version = "1.0.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cdb031dd78e28731d87d56cc8ffef4a8f36ca26c38fe2de700543e627f8a464a"

[[package]]
name = "chrono"
version = "0.4.19"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "670ad68c9088c2a963aaa298cb369688cf3f9465ce5e2d4ca10e6e0098a1ce73"
dependencies = [
 "libc",
 "num-integer",
 "num-traits",
 "time",
 "winapi",
]

[[package]]
name = "ctor"
version = "0.1.20"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5e98e2ad1a782e33928b96fc3948e7c355e5af34ba4de7670fe8bac2a3b2006d"
dependencies = [
 "quote",
 "syn",
]

[[package]]
name = "darling"
version = "0.13.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "757c0ded2af11d8e739c4daea1ac623dd1624b06c844cf3f5a39f1bdbd99bb12"
dependencies = [
 "darling_core",
 "darling_macro",
]

[[package]]
name = "darling_core"
version = "0.13.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2c34d8efb62d0c2d7f60ece80f75e5c63c1588ba68032740494b0b9a996466e3"
dependencies = [
 "fnv",
 "ident_case",
 "proc-macro2",
 "quote",
 "strsim",
 "syn",
]

[[package]]
name = "darling_macro"
version = "0.13.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ade7bff147130fe5e6d39f089c6bd49ec0250f35d70b2eebf72afdfc919f15cc"
dependencies = [
 "darling_core",
 "quote",
 "syn",
]

[[package]]
name = "difference"
version = "2.0.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "524cbf6897b527295dff137cec09ecf3a05f4fddffd7dfcd1585403449e74198"

[[package]]
name = "doku"
version = "0.9.0"
dependencies = [
 "chrono",
 "difference",
 "doku-derive",
 "indoc",
 "inventory",
 "paste",
 "pretty_assertions",
 "serde",
]

[[package]]
name = "doku-derive"
version = "0.9.0"
dependencies = [
 "darling",
 "proc-macro-error",
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "fnv"
version = "1.0.7"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3f9eec918d3f24069decb9af1554cad7c880e2da24a9afd88aca000531ab82c1"

[[package]]
name = "ghost"
version = "0.1.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1a5bcf1bbeab73aa4cf2fde60a846858dc036163c7c33bec309f8d17de785479"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "ident_case"
version = "1.0.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b9e0384b61958566e926dc50660321d12159025e767c18e043daf26b70104c39"

[[package]]
name = "indoc"
version = "0.3.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "47741a8bc60fb26eb8d6e0238bbb26d8575ff623fdc97b1a2c00c050b9684ed8"
dependencies = [
 "indoc-impl",
 "proc-macro-hack",
]

[[package]]
name = "indoc-impl"
version = "0.3.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ce046d161f000fffde5f432a0d034d0341dc152643b2598ed5bfce44c4f3a8f0"
dependencies = [
 "proc-macro-hack",
 "proc-macro2",
 "quote",
 "syn",
 "unindent",
]

[[package]]
name = "inventory"
version = "0.1.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0f0f7efb804ec95e33db9ad49e4252f049e37e8b0a4652e3cd61f7999f2eff7f"
dependencies = [
 "ctor",
 "ghost",
 "inventory-impl",
]

[[package]]
name = "inventory-impl"
version = "0.1.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "75c094e94816723ab936484666968f5b58060492e880f3c8d00489a1e244fa51"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "libc"
version = "0.2.98"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "320cfe77175da3a483efed4bc0adc1968ca050b098ce4f2f1c13a56626128790"

[[package]]
name = "num-integer"
version = "0.1.44"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d2cc698a63b549a70bc047073d2949cce27cd1c7b0a4a862d08a8031bc2801db"
dependencies = [
 "autocfg",
 "num-traits",
]

[[package]]
name = "num-traits"
version = "0.2.14"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9a64b1ec5cda2586e284722486d802acf1f7dbdc623e2bfc57e65ca1cd099290"
dependencies = [
 "autocfg",
]

[[package]]
name = "output_vt100"
version = "0.1.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "53cdc5b785b7a58c5aad8216b3dfa114df64b0b06ae6e1501cef91df2fbdf8f9"
dependencies = [
 "winapi",
]

[[package]]
name = "paste"
version = "0.1.18"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "45ca20c77d80be666aef2b45486da86238fabe33e38306bd3118fe4af33fa880"
dependencies = [
 "paste-impl",
 "proc-macro-hack",
]

[[package]]
name = "paste-impl"
version = "0.1.18"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d95a7db200b97ef370c8e6de0088252f7e0dfff7d047a28528e47456c0fc98b6"
dependencies = [
 "proc-macro-hack",
]

[[package]]
name = "pretty_assertions"
version = "0.6.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3f81e1644e1b54f5a68959a29aa86cde704219254669da328ecfdf6a1f09d427"
dependencies = [
 "ansi_term",
 "ctor",
 "difference",
 "output_vt100",
]

[[package]]
name = "proc-macro-error"
version = "1.0.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "da25490ff9892aab3fcf7c36f08cfb902dd3e71ca0f9f9517bea02a73a5ce38c"
dependencies = [
 "proc-macro-error-attr",
 "proc-macro2",
 "quote",
 "syn",
 "version_check",
]

[[package]]
name = "proc-macro-error-attr"
version = "1.0.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a1be40180e52ecc98ad80b184934baf3d0d29f979574e439af5a55274b35f869"
dependencies = [
 "proc-macro2",
 "quote",
 "version_check",
]

[[package]]
name = "proc-macro-hack"
version = "0.5.19"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "dbf0c48bc1d91375ae5c3cd81e3722dff1abcf81a30960240640d223f59fe0e5"

[[package]]
name = "proc-macro2"
version = "1.0.27"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f0d8caf72986c1a598726adc988bb5984792ef84f5ee5aa50209145ee8077038"
dependencies = [
 "unicode-xid",
]

[[package]]
name = "quote"
version = "1.0.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c3d0b9745dc2debf507c8422de05d7226cc1f0644216dfdfead988f9b1ab32a7"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "serde"
version = "1.0.126"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ec7505abeacaec74ae4778d9d9328fe5a5d04253220a85c4ee022239fc996d03"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.126"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "963a7dbc9895aeac7ac90e74f34a5d5261828f79df35cbed41e10189d3804d43"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "strsim"
version = "0.10.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "73473c0e59e6d5812c5dfe2a064a6444949f089e20eec9a2e5506596494e4623"

[[package]]
name = "syn"
version = "1.0.73"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f71489ff30030d2ae598524f61326b902466f72a0fb1a8564c001cc63425bcc7"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-xid",
]

[[package]]
name = "time"
version = "0.1.44"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6db9e6914ab8b1ae1c260a4ae7a49b6c5611b40328a735b21862567685e73255"
dependencies = [
 "libc",
 "wasi",
 "winapi",
]

[[package]]
name = "unicode-xid"
version = "0.2.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8ccb82d61f80a663efe1f787a51b16b5a51e3314d6ac365b08639f52387b33f3"

[[package]]
name = "unindent"
version = "0.1.7"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f14ee04d9415b52b3aeab06258a3f07093182b88ba0f9b8d203f211a7a7d41c7"

[[package]]
name = "version_check"
version = "0.9.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5fecdca9a5291cc2b8dcf7dc02453fee791a280f3743cb0905f8822ae463b3fe"

[[package]]
name = "wasi"
version = "0.10.0+wasi-snapshot-preview1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1a143597ca7c7793eff794def352d41792a93c481eb1042423ff7ff72ba2c31f"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
