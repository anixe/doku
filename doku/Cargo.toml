[package]
name = "doku"
version = "0.2.0-alpha.1"
authors = ["Athanasios Katsios <aka@anixe.io>", "Patryk Wychowaniec <patryk.wychowaniec@anixe.pl>"]
license = "MIT"
edition = "2018"
description = "A framework for building documentation"
readme = "../README.md"
repository = "https://github.com/anixe/doku"
keywords = ["doc", "docs", "documentation", "serialization"]

[dependencies]
chrono = { version = "0.4", optional = true }
doku-derive = { version = "=0.2.0-alpha.1", path = "../doku-derive" }
serde = "1.0"
serde_json = "1.0"
url = { version = "2", optional = true }

[features]
chrono-04 = ["chrono"]
url-2 = ["url"]

[dev-dependencies]
difference = "2.0"
indoc = "1.0"
once_cell = "<=1.14" # once_cell 1.15 uses rust edition 2021, which requires a greater rustc version
serde = { version = "1.0", features = ["derive"] }
test-case = "1.2"
trybuild = { version = "1.0", features = ["diff"] }
