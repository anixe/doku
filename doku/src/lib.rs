#![feature(crate_visibility_modifier)]

mod objects;
mod printers;

pub mod prelude {
    pub use doku_derive::*;
}

pub use self::{objects::*, printers::*};

pub fn to_json<Ty: ty::Provider>() -> String {
    JsonPrinter::new().print(&Ty::ty())
}

pub fn to_json_val<Ty: ty::Provider>(_: &Ty) -> String {
    to_json::<Ty>()
}
