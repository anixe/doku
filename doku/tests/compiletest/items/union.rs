use doku::Document;

#[derive(Document)]
union Person {
    age: u8,
}

fn main() {
    //
}
