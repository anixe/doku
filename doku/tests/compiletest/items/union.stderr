error: unions are not supported
 --> tests/compiletest/items/union.rs:4:7
  |
4 | union Person {
  |       ^^^^^^
